stdlib: Extend `AbstractBoard` pre_instantiation functionality

BobbyRBruce opened the pull request:
* Deprecates the setting of FS/SE mode via the `Simulator` module.
* Moved the creation of the `Root` object from the `Simulator` to the board.
* Moved the setting of `sim_quantum` from the `Simulator` to the processor.
* Allows for easier development of boards which support both SE and FS mode simulation by moving board setup function calls to occur after the set_workload function is call which sets a boards stats `is_fs` status.

BobbyRBruce committed:
stdlib: Deprecate Simulator 'full_system' param

THis is deprecated in favor of the board determining whether the
simulation is FS or SE. Usually this will be contingent on which
`set_workload` funciton has been called. Regardless, it is the board's
responsibility. The user should not need to explicitly declare this any
longer.

BobbyRBruce committed:
stdlib: Move Root obj creation from Simulator to Board

It makes much more sense for the Root Object to be create within the
board and passed where required. Creating it in the Simulator class is
not required.

For this to work the signuature of the `_pre_instantiate` function in
`AbstractBoard` has been updated to return the Root object.

BobbyRBruce committed:
stdlib: Move 'sim_quantum' set from Simulator to Processor

The setting of the `sim_quantum` parameter makes considerably more sense
to occur in the Processor. Through the `_pre_instnatiate` functions this
is now possible.

BobbyRBruce commented:
((Note: This is still a draft but I wanted to run the tests so have made it "ready for review" purely for this reason)).

BobbyRBruce commented:
@mahyarsamani : Is this ok with you?

powerjg commented on src/python/gem5/components/boards/arm_board.py:
```diff
@@ -327,8 +328,8 @@ def connect_system_port(self, port: Port) -> None:
     @overrides(AbstractBoard)
-    def _pre_instantiate(self):
-        super()._pre_instantiate()
+    def _pre_instantiate(self, full_system: Optional[bool]) -> None:
```
I don't think `full_system` is "optional" here since it doesn't have a default.

powerjg commented on src/python/gem5/components/boards/riscv_board.py:
```diff
@@ -498,7 +501,7 @@ def get_disk_device(self):
 
     @overrides(AbstractSystemBoard)
-    def _pre_instantiate(self):
+    def _pre_instantiate(self, full_system: Optional[bool]):
```
I don't think `full_system` is "optional" here since it doesn't have a default.

powerjg commented on src/python/gem5/prebuilt/riscvmatched/riscvmatched_board.py:
```diff
@@ -313,7 +313,7 @@ def _setup_memory_ranges(self):
 
     @overrides(AbstractSystemBoard)
-    def _pre_instantiate(self):
+    def _pre_instantiate(self, full_system: Optional[bool]) -> None:
```
I don't think `full_system` is "optional" here since it doesn't have a default.

powerjg commented on src/python/gem5/components/boards/abstract_board.py:
```diff
@@ -384,13 +385,40 @@ def _post_instantiate(self):
+
+        # 3. Call any of the components' `_pre_instantiate` functions.
+        # Right now, only the processor requires this.
+        self.get_processor()._pre_instantiate(root)
```
Let's go ahead and add this for the `cache_hierarchy` and `memory_system`. I need the cache hierarchy to get the power model example in the bootcamp to work.

BobbyRBruce replied:
Unless I'm very confused I believe I already have. Perhaps you were looking at an old commit? The next two lines call the cache hierarchy and memory system pre_instantiate functions

BobbyRBruce replied:
Disregard, that... I was confused. I think I deleted some god reverting your changes. i'll add these

BobbyRBruce replied:
Done

powerjg reviewed (commented):
A couple of small things below. I pushed a commit to fix these things.

powerjg commented:
Apologies for messing up anything with the force push. I was trying to make your life easier, but pre-commit is broken on my system after upgrading to 24.04.

BobbyRBruce committed:
stdlib: Add `_pre_instantiate` funcs for caches and memory

Note: At present this is not used but these functions can be filled
or overriden in subclasses as required.

BobbyRBruce reviewed (commented):
@powerjg : I take it I can remove the last commit in this chain?  https://github.com/gem5/gem5/pull/1497/commits/14e12907346e4e9132c7bf969c0634f8115e5c2c. It's only purpose was to demo this stuff to Mayhar. I'm not sure it's actually useful

powerjg commented:
> @powerjg : I take it I can remove the last commit in this chain? https://github.com/gem5/gem5/commit/14e12907346e4e9132c7bf969c0634f8115e5c2c. It's only purpose was to demo this stuff to Mayhar. I'm not sure it's actually useful

It's up to you. We can remove it and let Mahyar push a change if he thinks it's important.

powerjg commented on src/python/gem5/components/memory/abstract_memory_system.py:
```diff
@@ -50,6 +51,21 @@ class AbstractMemorySystem(SubSystem):
+        """
+        pass
+
+    def _pre_instantiate(self, root: Root) -> None:
+        """Called to set up anything needed before ``m5.instantiate``."""
```
I think this should be removed.

BobbyRBruce replied:
Yeah, that was a mistake. Removed.

powerjg reviewed (commented):
One little thing. Sorry if I made things harder for you by adding that commit.

powerjg commented:
I suggest dropping the branch predictor from this pr and creating another with an example.

powerjg reviewed (changes requested):
The branch predictor part of this (most likely) doesn't work. Let's drop that commit and then this can be merged. I would like this merged ASAP because we have other changes that depend on this.

BobbyRBruce committed:
stdlib: Mv setup_board/setup_mem_ranges calls to set_fs

This change allows for the `_setup_memory_range` and `_setup_board`
functions to know if the board is to run a FS or SE workload, thus
allowing for a baord to handle both cases considerably easier than
before. With this change all functions are called after FS or SE
is declared via the `_set_fullsystem` function and thus all can
accomodate for SE and FS workloads.

BobbyRBruce commented:
> The branch predictor part of this (most likely) doesn't work. Let's drop that commit and then this can be merged. I would like this merged ASAP because we have other changes that depend on this.

I've removed it.

I've added one more patch to this chain. It moves board setup functions to occur after it's decided whether the board is in FS or SE (which is decided based on what set_workload). This allows for easier implementation of boards that support both. All setup functions can call `is_fullsystem` and modify the simulated system accordingly. Very little now happens before the workloads are specified.

Harshil2107 reviewed (approved):
I tested this change on top of the arm demo board and the boot fs simulation ran successfully. Overall the other changes also look good to me.
